application/common/components/MfaBackendWebAuthn.php

```diff
@@ -125,29 +135,65 @@ public function verify(int $mfaId, $value, string $rpOrigin = '')
             $mfa->external_uuid
         );
 
+
         if (!is_array($value)) {
             $value = Json::decode($value);
             if ($value == null) {
                 throw new ServerErrorHttpException("Missing data or unable to decode as JSON", 1638447364);
             }
         }
 
-        if ($mfa->verified === 1) {
-            return $this->client->webauthnValidateAuthentication($headers, $value);
-        } else {
-            $results = $this->client->webauthnValidateRegistration($headers, $value);
-            if (isset($results['key_handle_hash'])) {
-                $mfa->verified = 1;
-                $mfa->key_handle_hash = $results['key_handle_hash'];
-                if (! $mfa->save()) {
-                    throw new ServerErrorHttpException(
-                        "Unable to save WebAuthn record after verification. Error: " . print_r($mfa->getFirstErrors(), true)
-                    );
-                }
-                return true;
+        // If the verifyType is not registration, finish the login process.
+        if ($verifyType != Mfa::VERIFY_REGISTRATION) {
+            // Ensure there is at least one verified MfaWebauthn for that Mfa
+            $webauthnCount = MfaWebauthn::find()->where(['mfa_id' => $mfa->id,'verified' => 1])->count();
+            if ($webauthnCount < 1) {
+                throw new NotFoundHttpException("No verified MFA Webauthn record found for MFA ID: " . $mfa->id, 1659637860);
             }
+            return $this->client->webauthnValidateAuthentication($headers, $value);
+        }
+
+        // Assume a new webauthn was requested and finish its registration process
+        $results = $this->client->webauthnValidateRegistration($headers, $value);
+
+        if (! isset($results['key_handle_hash'])) {
             return false;
         }
+
+        $label = $mfa->getReadableType();
+        self::createWebauthn($mfa, $results['key_handle_hash']);
+
+        return true;
+    }
+
+
+
+    /**
+     * @param string $label
+     * @param string $keyHandleHash
+     * @return void
+     * @throws ServerErrorHttpException
+     */
+    public static function createWebauthn(Mfa $mfa, string $keyHandleHash): void {
+        if ($mfa->type != self::TYPE_WEBAUTHN) {
+            return;
+        }
```

Comment:
Counter to some of my earlier suggestions (😉), you will want to keep this sanity check when you move this back to MfaWebauthn, since it will need to be public. However you could change this to throw an exception, so that if we do later accidentally call this in some non-WebAuthn context, we'll know about that mistake as soon as we make it.